iup/__init__.py

```diff
@@ -4,77 +4,45 @@
 import numpy as np
 import abc
 from typing_extensions import Self
-from typing import List
-import re
+from typing import Sequence
 
 
-class ValidateData(pl.DataFrame, metaclass=abc.ABCMeta):
+class Data(pl.DataFrame):
     """
     Abstract class for observed data and forecast data.
     """
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-
-    @abc.abstractmethod
-    def validate(self) -> None:
-        pass
-
-
-class UptakeData(ValidateData):
-    """
-    Abstract class for different forms of uptake data.
-    """
-
-    def __init__(self, *args, **kwargs):
-        """
-        Initialize an UptakeData object as a polars data frame plus validation.
-        """
-        super().__init__(*args, **kwargs)
+        self.validate()
 
     def validate(self):
-        """
-        Validate that an UptakeData object has the two key columns:
-        date and uptake estimate (% of population). There may be others.
-        """
+        raise NotImplementedError("Subclasses must implement this method.")
 
-        self.assert_columns_found(["date", "estimate"])
-        self.assert_columns_type(["date"], pl.Date)
-        self.assert_columns_type(["estimate"], pl.Float64)
+    def assert_in_schema(self, names_types: dict[str, pl.DataType]):
+        """Verify that column of the expected types are present in the data frame
 
-    def assert_columns_found(self, columns: List[str]):
+        Args:
+            names_types (dict[str, pl.DataType]): Column names and types
         """
-        Verify that expected columns are found.
+        for name, type_ in names_types.items():
+            if name not in self.schema.names():
+                raise RuntimeError(f"Column '{name}' not found")
+            elif (name, type_) not in self.schema.items():
+                actual_type = self.schema.to_python()[name]
+                f"Column '{name}' has type {actual_type}, not {type_}"
 
-        Parameters
-        columns: List[str]
-            names of expected columns
-        """
-        for col in columns:
-            assert col in self.columns, f"Column {col} is expected but not found."
+            assert (name, type_) in self.schema.items()
 
-    def assert_columns_type(self, columns: List[str], dtype):
-        """
-        Verify that columns have the expected type.
 
-        Parameters
-        columns: List[str]
-            names of columns for which to check type
-        dtype:
-            data type for each listed column
-        """
-        for col in columns:
-            assert (
-                self[col].dtype == dtype
-            ), f"Column {col} should be {dtype} but is {self[col].dtype}"
-
-    def with_columns(self, *args, **kwargs):
+class UptakeData(Data):
```

Comment:
This class is named "UptakeData," but below there are many spots where "ValidatedUptake" is used (often directly replacing "UptakeData." Looks like we must decide which of these terms we like better. I prefer "UptakeData" because "Validated" seems unnecessary - we know we are going to validate whenever we create one of these objects.

Reply:
The `UptakeData` was replaced with `ValidatedData` in PR#64, while this branch is opened before PR#64 merged to main, the original `UptakeData` is reserved. Then `ValidatedData` is changed to `ValidatedUptake` based on this [comment](https://github.com/CDCgov/cfa-immunization-uptake-projection/pull/64#discussion_r1867830878).

Reply:
Makes sense as to how we got here! But we still must decide among these names, and my vote still for "UptakeData" as opposed to anything with "Validated" in it.

Reply:
I'm fine using `UptakeData` to include observed data (`IncidentUptakeData`) and predictions ('QuantileForecast`) since they are both uptake data set. Let us know your thought! @swo

Reply:
I just completed a rebase; I called the top level class just `Data`. Open to changes.